### ISSUE TITLE
Vulnerable packages and or dependencies found in keycloak 21.0.1 quarkus distribution

### ISSUE THREAD

**Issue by ahrycej** [keycloak/keycloak#19080]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

dist/quarkus

### Describe the bug

anchore enterprise finds following packages and or dependecies in keyckloak:

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:sshd-common
https://github.com/advisories/GHSA-fhw8-8j55-vwgq

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:jackson-databind
https://github.com/advisories/GHSA-jjjh-jjxp-wpff

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:wildfly-elytron
https://github.com/advisories/GHSA-jmj6-p2j9-68cp

/opt/keycloak/lib/lib/main/org.yaml.snakeyaml-1.33.jar:snakeyaml
https://github.com/advisories/GHSA-mjmj-j48q-9wg2

/opt/keycloak/lib/lib/main/io.vertx.vertx-web-4.3.4.jar:vertx-web
https://github.com/advisories/GHSA-53jx-vvf9-4x38

### Version

21.0.1

### Expected behavior

ssd-common is 2.7.0 should be at least 2.9.2
jackson databind is 2.13.3 should be at least 2.13.4.1
wildfly-elytron is 1.20.1,Final should be at least 1.20.3.Final
snakeyaml is 1.33 should be 2.0
vertx-web is 4.3.4 should be at least 4.3.8

### Actual behavior

exploitable vulnerabilities present
ssd-common is 2.7.0
jackson databind is 2.13.3 
wildfly-elytron is 1.20.1,Final 
snakeyaml is 1.33
vertx-web is 4.3.4 

### How to Reproduce?

build docker image with keycloak 21.0.1
run grype and or anchore enterprise

### Anything else?

if some of this vulnerabilities are not exploitable would be good to know why

**Comment by kkcmadhu** [comment]
Yup.. my security audit team too raised concern with critical vulnarablities in ssd-common 2.7.x  :(

**Comment by ahrycej** [comment]
There were many more issues in previous varsions, there is a great progress, still few issues left to resolve.

**Comment by alapierre** [comment]
most important is to update `snakeyaml 1.33` to 2.0 because `CVE-2022-1471` is marked as `Critical` by many security tools. It will by nice if `json-smart` can be updated into 2.4.9, too. 

![image](https://user-images.githubusercontent.com/5985325/230033058-6bb4d449-aee6-4966-a59a-2cf319563003.png)

**Comment by ahrycej** [comment]
I believe updating wildfly-elytron to  1.20.3.Final shall resolve jackson-databind dependancy, I will trigger new scan I verify location of json-smart, I could not see this but my scan is in a way old

**Comment by jonkoops** [comment]
@vmuzikar any chance we could somehow automate these kinds of patches? It seems like it will take up a lot of work doing this manually in a reactive manner all the time.

**Comment by alapierre** [comment]
there is nice tool called Mend Renovate - it is free for open source projects. It can create pull requests with dependency updates and security patches. It is full integrated with GitHub - I'm using it with my projects.

**Comment by ahrycej** [comment]
I would say if you integrate grype to the pipline and start gating it, then we shall have all fixes always in. Policy shall be zero vulnerability with published fix

**Comment by ahrycej** [comment]
We have 21.0.1 quarkus distribution and my tools dont find json-smart using anchore enterprise (grype/syft). will re-confirm when we build image with 21.0.2

**Comment by alapierre** [comment]
security tools have different algorithms, eg. Trivy by Aquasec found this in Java dependencies
![image](https://user-images.githubusercontent.com/5985325/230072607-bf6ce976-59e6-4f0f-aae0-51d330adb0c5.png)

and this in docker image

![image](https://user-images.githubusercontent.com/5985325/230072811-03deddb4-e80f-46c7-bd33-4e2186d333ed.png)

**Comment by ahrycej** [comment]
I get it, so it is java related, need to check it closer, we have still java 11

**Comment by ahrycej** [comment]
Is this the same?
https://github.com/microsoft/ApplicationInsights-Java/issues/2982

**Comment by ahrycej** [comment]
on your screenshots I cant see CVE-2023-1370 for json-smart, could not find any information regarding java dependancy either, still looking

**Comment by alapierre** [comment]
Yes, one of my my images adds Azure applicationinsights-agent and this is why Aquasec finds `CVE-2023-1370` in json-smart and other tools not. Only for this image I'm using Aquasec scanner.

**Comment by alapierre** [comment]
according to `xstream`, if ypu look here: https://mvnrepository.com/artifact/com.thoughtworks.xstream/xstream/1.4.20

You can confirm, that `CVE-2022-40152` is applicable for version 1.4.20. Security tools read java dependency vulnerabilities  from maven repositories.

**Comment by JanEgner** [comment]
fwiw - the keycloak 21.0.2 quarkus distribution still contains a fairly significant number of wildfly-elytron submodules with version 1.20.1.Final (i.e., CVE-2022-3143 is still being flagged). I suspect that quarkus' dependency management gets confused by the many conflicting versions of elytron pulled in by transitive dependencies.
Reproducible by building the distro from the current `release/21.0` branch and looking at the resulting zip archive, in path `lib/lib/main`:
![image](https://user-images.githubusercontent.com/24408313/232467833-03deb0b3-291a-4578-a8b9-e04c5716c829.png)

It seems that hard-coding the wildfly-elytron-* dependencies with the correct version tag in `quarkus/pom.xml` helps, but that doesn't sound like a correct approach.

**Comment by abstractj** [comment]
Thank you for bringing these issues to our attention and reporting. Our team is reviewing the number of third-party dependencies and preparing updates for the upcoming releases. Please find below an explanation that aims to bring more context to the conversation. That doesn't mean that our team won't address those issues.

- sshd-common (CVE-2022-45047): Keycloak does not use OpenSSH, even though it is impacted, the code is not vulnerable.
- WildFly Elytron (CVE-2022-3143): From my understanding, https://github.com/advisories/GHSA-jmj6-p2j9-68cp affects the following packages: [security.snyk.io/vuln/maven?search=CVE-2022-3143](https://security.snyk.io/vuln/maven?search=CVE-2022-3143). All those packages were bumped to 1.20.1, so I believe we should be safe if we consider that https://github.com/advisories/GHSA-jmj6-p2j9-68cp impacts versions equal to, and less than 1.15.15 of Elytron.
- SnakeYAML: as described in the CVE "SnakeYaml's Constructor() class does not restrict types which can be instantiated during deserialization". Keycloak does not make any usage of SnakeYaml constructor and uap-java uses SafeConstructor ([ua-parser/uap-java@50535db/src/main/java/ua_parser/Parser.java#L26](https://github.com/ua-parser/uap-java/blob/50535dbc02953ccf4614006e26ace52af1c3b8b4/src/main/java/ua_parser/Parser.java#L26)). For this reason, the priority is considered low, even though both of the projects are impacted, but not vulnerable.
- vertx-web (CVE-2023-24815): reported on https://github.com/keycloak/keycloak/issues/17476. Keycloak does not make any use of `StaticHandle` from vertx-web
- jackson-databind:
	-  (CVE-2022-42003): Keycloak does not enable `UNWRAP_SINGLE_VALUE_ARRAYS` feature
	- ( CVE-2022-42004): we do not make any usage of `BeanDeserializer._deserializeFromArray`, or provide custom choices for deserialization

**Comment by JanEgner** [comment]
Thanks a lot @abstractj, very helpful! I just want to point out that the Wildfly Elytron issue seems to still be present in Elytron 1.20.1 - the GitHub advisory https://github.com/advisories/GHSA-jmj6-p2j9-68cp says that versions `>= 1.16.0.CR1, < 1.20.3.Final ` are affected as well as the older line, unfortunately.

**Comment by ahrycej** [comment]
I can confirm that following vulnerabilities are not present already in 21.1.1
Thank you!

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:sshd-common
https://github.com/advisories/GHSA-fhw8-8j55-vwgq

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:jackson-databind
https://github.com/advisories/GHSA-jjjh-jjxp-wpff

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:wildfly-elytron

**Comment by JanEgner** [comment]
@ahrycej - while I agree that the first two vulnerabilities you list do not seem to affect Keycloak 21.1.1 given abstractj's clarifications, I still stand by my statement from a month ago that the third issue _is_ still present in 21.1.1.
The vulnerability might not be exploitable in the context of Keycloak, but I did not find any indications confirming this assumption.

**Comment by ahrycej** [comment]
Sorry yes you are right, I was checking only for presense of vulnerable sshd
